pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,435 @@
+
+    Attributes
+    ----------
+    client: :class:`Client`
+        The client object
+    managers: Dict[:class:`str`, :class:`~typing.Any`]
+        Dictionary of managers
+    register: Dict[:class:`str`, :class:`~pincer.objects.app.command.InteractableStructure`]
+        Dictionary of ``InteractableStructure``
+    built_register: Dict[:class:`str`, :class:`~pincer.objects.app.command.InteractableStructure`]
+        Dictionary of ``InteractableStructure`` where the commands are converted to
+        the format that Discord expects for sub commands and sub command groups.
+    """  # noqa: E501
+
+    has_been_initialized = False
+    managers: List[Interactable] = []
+    register: Dict[str, InteractableStructure] = {}
+    built_register: Dict[str, AppCommand] = {}
+
+    # Endpoints:
+    __get = "/commands"
+    __delete = "/commands/{command.id}"
+    __update = "/commands/{command.id}"
+    __add = "/commands"
+    __add_guild = "/guilds/{command.guild_id}/commands"
+    __get_guild = "/guilds/{guild_id}/commands"
+    __update_guild = "/guilds/{command.guild_id}/commands/{command.id}"
+    __delete_guild = "/guilds/{command.guild_id}/commands/{command.id}"
+
+    def __init__(self, client: Client):
+        self.client = client
+        self._api_commands: List[AppCommand] = []
+        _log.debug(
+            "%i commands registered.", len(ChatCommandHandler.register.items())
+        )
+
+        self.__prefix = f"applications/{self.client.bot.id}"
+
+    async def get_commands(self) -> List[AppCommand]:
+        """|coro|
+
+        Get a list of app commands from Discord
+
+        Returns
+        -------
+        List[:class:`~pincer.objects.app.command.AppCommand`]
+            List of commands.
+        """
+        # TODO: Update if discord adds bulk get guild commands
+        guild_commands = await gather(*(
+            self.client.http.get(
+                self.__prefix + self.__get_guild.format(
+                    guild_id=guild.id if isinstance(guild, Guild) else guild
+                )
+            ) for guild in self.client.guilds
+        ))
+        return list(
+            map(
+                AppCommand.from_dict,
+                await self.client.http.get(self.__prefix + self.__get)
+                + [cmd for guild in guild_commands for cmd in guild],
+            )
+        )
+
+    async def remove_command(self, cmd: AppCommand):
+        """|coro|
+
+        Remove a specific command
+
+        Parameters
+        ----------
+        cmd : :class:`~pincer.objects.app.command.AppCommand`
+            What command to delete
+        """
+        # TODO: Update if discord adds bulk delete commands
+        if cmd.guild_id:
+            _log.info(
+                "Removing command `%s` with guild id %d from Discord",
+                cmd.name,
+                cmd.guild_id,
+            )
+        else:
+            _log.info("Removing global command `%s` from Discord", cmd.name)
```

Comment:
Shouldn't we warn the user that this process might take a long time?

Reply:
not a problem for this PR